File: benches/bigdecimal/python/bench.py
#!/usr/bin/env python3
"""Cross-language BigDecimal benchmark — Python (decimal.Decimal) side.

Reads cases/<op>.toml, expands `{C,N}` patterns, auto-tunes iteration count
to ~50ms per case, and emits one CSV per case to logs/python_<op>_<ts>.csv:

    timestamp,language,op,case_name,result,ns_per_iter

Python's decimal.Decimal is the **oracle**: the aggregator marks the `match`
column as OK iff every other language's result is numerically equal to
Python's, so formatting-only differences such as trailing zeros do not
count as DIFFs.

Usage:
    python3 bench.py --op multiply --cases-dir ../cases --logs-dir ../logs
"""

from __future__ import annotations

import argparse
import csv
import decimal
import os
import re
import sys
import time
from datetime import datetime, timezone

try:
    import tomllib  # py 3.11+
except ImportError:
    import tomli as tomllib  # type: ignore


PATTERN_RE = re.compile(r"\{([^{}]*),(\d+)\}")


def expand(s: str) -> str:
    """Expand `{C,N}` repeat patterns. Last comma wins (matches Mojo)."""
    out = []
    i = 0
    while i < len(s):
        if s[i] == "{":
            close = s.find("}", i + 1)
            if close < 0:
                out.append(s[i])
                i += 1
                continue
            inner = s[i + 1 : close]
            comma = inner.rfind(",")
            if comma < 0:
                out.append(s[i : close + 1])
                i = close + 1
                continue
            payload = inner[:comma]
            try:
                n = int(inner[comma + 1 :])
            except ValueError:
                out.append(s[i : close + 1])
                i = close + 1
                continue
            out.append(payload * n)
            i = close + 1
        else:
            out.append(s[i])
            i += 1
    return "".join(out)


# ----- per-op kernels ---------------------------------------------------


def _round_args(b: str):
    ndigits_str, mode = b.split("|", 1)
    py_mode = {
        "ROUND_DOWN": decimal.ROUND_DOWN,
        "ROUND_UP": decimal.ROUND_UP,
        "ROUND_HALF_UP": decimal.ROUND_HALF_UP,
        "ROUND_HALF_DOWN": decimal.ROUND_HALF_DOWN,
        "ROUND_HALF_EVEN": decimal.ROUND_HALF_EVEN,
        "ROUND_CEILING": decimal.ROUND_CEILING,
        "ROUND_FLOOR": decimal.ROUND_FLOOR,
    }[mode]
    return int(ndigits_str), py_mode


def _fmt(d) -> str:
    """Format a Decimal as fixed-point (no scientific notation).

    For non-Decimal values (e.g. comparison int, str), fall back to str().
    Important: do NOT apply `+d`. The arithmetic ops (`a+b`, `a*b`, ...)
    already round their result to the active context precision; calling
    `+d` afterwards is at best idempotent for arithmetic ops but adds an
    unwanted rounding pass for `from_string`/`to_string` (where the
    timed kernel deliberately preserves the input digits). Keeping the
    display path rounding-free keeps the displayed result aligned with
    what the kernel actually computed.
    """
    if isinstance(d, decimal.Decimal):
        return format(d, "f")
    return str(d)


def make_kernel(op: str, a: str, b: str, precision: int):
    """Return (display_result_str, kernel_callable_no_args, supported_bool)."""
    decimal.getcontext().prec = precision
    da = decimal.Decimal(a)
    # `b` for the `round` op encodes "ndigits|MODE" (not a Decimal).
    if op == "round" or b in ("", None):
        db = decimal.Decimal(0)
    else:
        db = decimal.Decimal(b)

    if op == "add":
        return _fmt(da + db), (lambda: da + db), True
    if op == "subtract":
        return _fmt(da - db), (lambda: da - db), True
    if op == "multiply":
        return _fmt(da * db), (lambda: da * db), True
    if op == "divide":
        return _fmt(da / db), (lambda: da / db), True
    if op == "comparison":
        cmp = -1 if da < db else (1 if da > db else 0)
        return str(cmp), (lambda: -1 if da < db else (1 if da > db else 0)), True
    if op == "from_string":
        # Parsing is precision-insensitive: do not apply current-context
        # rounding. The kernel times exactly the parse path.
        return _fmt(decimal.Decimal(a)), (lambda: decimal.Decimal(a)), True
    if op == "to_string":
        return _fmt(da), (lambda: format(da, "f")), True
    if op == "sqrt":
        return _fmt(da.sqrt()), (lambda: da.sqrt()), True
    if op == "exp":
        return _fmt(da.exp()), (lambda: da.exp()), True
    if op == "ln":
        return _fmt(da.ln()), (lambda: da.ln()), True
    if op == "root":
        # Python decimal has no nth root; emulate via x**(1/n) for the
        # *result* (oracle) but mark as supported. n is bc.b.
        n = decimal.Decimal(b)
        recip = decimal.Decimal(1) / n
        result = da**recip
        return _fmt(result), (lambda: da**recip), True
    if op == "round":
        ndigits, py_mode = _round_args(b)
        decimal.getcontext().rounding = py_mode
        # Python's Decimal.__round__(n) uses current context rounding mode.
        return _fmt(round(da, ndigits)), (lambda: round(da, ndigits)), True
    raise ValueError(f"unknown op: {op}")


# ----- timing -----------------------------------------------------------

TARGET_NS = 50_000_000
MIN_RES_NS = 100_000  # 100µs floor per rep for resolution
MAX_WALL_NS = 500_000_000  # 500ms total wall per case


def bench_kernel(kernel, iter_hint: int, precision: int = 100) -> float:
    """Return best-of-N ns/iter, auto-tuned.

    Strategy:
      - Target ~50 ms per rep (TARGET_NS).
      - Resolution floor: ensure each rep takes ≥100 µs so cheap ops do not
        collapse to 0 ns/iter at high precision.
      - Adaptive `reps`: shrink from 3 → 1 for very-slow ops to bound wall
        time per case at ~500 ms.
    `precision` is currently informational; iter scaling derives from the
    measured per-iter cost rather than precision directly.
    """
    del precision  # not used: cal_ns drives the choice
    # calibrate
    t0 = time.perf_counter_ns()
    r = kernel()
    cal = time.perf_counter_ns() - t0
    if cal <= 0:
        cal = 1
    n = TARGET_NS // cal
    n_min_res = MIN_RES_NS // cal
    if n < n_min_res:
        n = n_min_res
    if n < 3:
        n = 3
    if n > iter_hint:
        n = iter_hint
    if n < 1:
        n = 1
    iters = int(n)
    per_rep = iters * cal
    reps = 3
    if per_rep > 0:
        reps = max(1, min(3, MAX_WALL_NS // per_rep))
    best = 1 << 62
    for _ in range(int(reps)):
        t0 = time.perf_counter_ns()
        for _ in range(iters):
            r = kernel()
        dt = time.perf_counter_ns() - t0
        if dt < best:
            best = dt
    _ = r
    return best / iters


# ----- main -------------------------------------------------------------


def main() -> int:
    ap = argparse.ArgumentParser()
    ap.add_argument("--op", required=True)
    ap.add_argument("--cases-dir", default="../cases")
    ap.add_argument("--logs-dir", default="../logs")
    ap.add_argument(
        "--precision",
        type=int,
        default=None,
        help="Override precision from cases TOML.",
    )
    args = ap.parse_args()

    toml_path = os.path.join(args.cases_dir, f"{args.op}.toml")
    with open(toml_path, "rb") as f:
        doc = tomllib.load(f)
    cfg = doc.get("config", {})
    iter_hint = int(cfg.get("iterations", 1000))
    precision = (
        int(args.precision)
        if args.precision is not None
        else int(cfg.get("precision", 28))
    )
    cases = doc.get("cases", [])

    os.makedirs(args.logs_dir, exist_ok=True)
    ts = datetime.now(timezone.utc).strftime("%Y%m%d_%H%M%S")
    log_path = os.path.join(args.logs_dir, f"python_{args.op}_p{precision}_{ts}.csv")
    print(f"# python decimal.Decimal {args.op} (prec={precision}, hint={iter_hint})")
    print(f"{'case':<44}{'result':<36}ns/iter")
    with open(log_path, "w", newline="") as fout:
        w = csv.writer(fout, lineterminator="\n")
        w.writerow(
            [
                "timestamp",
                "language",
                "op",
                "case_name",
                "result",
                "ns_per_iter",
                "precision",
            ]
        )
        for c in cases:
            name = c["name"]
            a = expand(c["a"])
            b = expand(c.get("b", "")) if c.get("b") not in (None, "") else ""
            try:
                result, kernel, _ = make_kernel(args.op, a, b, precision)
                per_ns = bench_kernel(kernel, iter_hint, precision)
            except Exception as exc:
                result = f"ERR: {exc.__class__.__name__}: {exc}"
                per_ns = 0.0
            short = result if len(result) <= 34 else result[:34]
            print(f"{name:<44}{short:<36}{per_ns:.2f}")
            w.writerow(
                [ts, "python", args.op, name, result, f"{per_ns:.4f}", str(precision)]
            )
    print(f"wrote {log_path}", file=sys.stderr)
    return 0


if __name__ == "__main__":
    raise SystemExit(main())

File: benches/bigdecimal/js/bench.mjs
// Cross-language BigDecimal benchmark — JavaScript (decimal.js) side.
//
// Reads cases/<op>.toml, expands {C,N} repeat patterns, auto-tunes iteration
// count to ~50ms per case, and emits one CSV per case to
// logs/js_<op>_<ts>.csv. Schema (mirrors mojo/python sides):
//
//     timestamp,language,op,case_name,result,ns_per_iter
//
// Usage:  node bench.mjs --op multiply --cases-dir ../cases --logs-dir ../logs

import { readFileSync, writeFileSync, mkdirSync, existsSync } from "node:fs";
import { dirname, join } from "node:path";
import { fileURLToPath } from "node:url";
import process from "node:process";
import { performance } from "node:perf_hooks";

import Decimal from "decimal.js";
import TOML from "@iarna/toml";

const TARGET_NS = 50_000_000n;

function expand(s) {
    // Mirror Mojo/Python {C,N} repeat expansion (last comma wins).
    let out = "";
    let i = 0;
    while (i < s.length) {
        if (s[i] === "{") {
            const close = s.indexOf("}", i + 1);
            if (close < 0) {
                out += s[i++];
                continue;
            }
            const inner = s.slice(i + 1, close);
            const comma = inner.lastIndexOf(",");
            if (comma < 0) {
                out += s.slice(i, close + 1);
                i = close + 1;
                continue;
            }
            const payload = inner.slice(0, comma);
            const n = parseInt(inner.slice(comma + 1), 10);
            if (Number.isNaN(n)) {
                out += s.slice(i, close + 1);
            } else {
                out += payload.repeat(n);
            }
            i = close + 1;
        } else {
            out += s[i++];
        }
    }
    return out;
}

function csvQuote(s) {
    if (typeof s !== "string") s = String(s);
    if (s.includes(",") || s.includes('"') || s.includes("\n") || s.includes("\r")) {
        return '"' + s.replaceAll('"', '""') + '"';
    }
    return s;
}

function ts() {
    // YYYYMMDD_HHMMSS in UTC (matches mojo/rust/go sides).
    const d = new Date();
    const pad = (n) => String(n).padStart(2, "0");
    return (
        d.getUTCFullYear() +
        pad(d.getUTCMonth() + 1) +
        pad(d.getUTCDate()) +
        "_" +
        pad(d.getUTCHours()) +
        pad(d.getUTCMinutes()) +
        pad(d.getUTCSeconds())
    );
}

function makeKernel(op, a, b, precision) {
    // decimal.js precision is sig digits, set globally. Set toExpPos/toExpNeg
    // wide so toString() prefers plain decimal form (closer to Python /
    // decimo's default formatting and avoids spurious "DIFF" rows).
    Decimal.set({ precision: precision, toExpPos: 999, toExpNeg: -999 });
    const da = new Decimal(a);
    const db = b === "" ? new Decimal(0) : new Decimal(b);

    switch (op) {
        case "add": return [da.plus(db).toString(), () => da.plus(db)];
        case "subtract": return [da.minus(db).toString(), () => da.minus(db)];
        case "multiply": return [da.times(db).toString(), () => da.times(db)];
        case "divide": return [da.div(db).toString(), () => da.div(db)];
        case "comparison": {
            const c = da.cmp(db);
            return [String(c), () => da.cmp(db)];
        }
        case "from_string": return [new Decimal(a).toString(), () => new Decimal(a)];
        case "to_string": return [da.toString(), () => da.toString()];
        case "sqrt": return [da.sqrt().toString(), () => da.sqrt()];
        case "exp": return [da.exp().toString(), () => da.exp()];
        case "ln": return [da.ln().toString(), () => da.ln()];
        case "root": {
            const recip = new Decimal(1).div(db);
            return [da.pow(recip).toString(), () => da.pow(recip)];
        }
        case "round": {
            const [ndigStr, mode] = b.split("|");
            const ndigits = parseInt(ndigStr, 10);
            const rmMap = {
                ROUND_DOWN: Decimal.ROUND_DOWN,        // toward zero
                ROUND_UP: Decimal.ROUND_UP,            // away from zero
                ROUND_HALF_UP: Decimal.ROUND_HALF_UP,
                ROUND_HALF_DOWN: Decimal.ROUND_HALF_DOWN,
                ROUND_HALF_EVEN: Decimal.ROUND_HALF_EVEN,
                ROUND_CEILING: Decimal.ROUND_CEIL,
                ROUND_FLOOR: Decimal.ROUND_FLOOR,
            };
            const rm = rmMap[mode];
            // toDecimalPlaces rounds to ndigits decimal places (negative ndigits
            // rounds the integer part, which decimal.js supports via toDP-then-multiply
            // workaround). decimal.js's toDecimalPlaces accepts negative dp.
            return [da.toDecimalPlaces(ndigits, rm).toString(),
            () => da.toDecimalPlaces(ndigits, rm)];
        }
        default: throw new Error("unknown op: " + op);
    }
}

function bench(kernel, iterHint) {
    // calibrate
    let t0 = process.hrtime.bigint();
    let r = kernel();
    let cal = process.hrtime.bigint() - t0;
    if (cal === 0n) cal = 1n;
    let iters = Number(TARGET_NS / cal);
    if (iters < 3) iters = 3;
    if (iters > iterHint) iters = iterHint;
    let best = 1n << 62n;
    for (let rep = 0; rep < 3; rep++) {
        t0 = process.hrtime.bigint();
        for (let i = 0; i < iters; i++) r = kernel();
        const dt = process.hrtime.bigint() - t0;
        if (dt < best) best = dt;
    }
    return Number(best) / iters;
}

function main() {
    let op = "add";
    let casesDir = "../cases";
    let logsDir = "../logs";
    const argv = process.argv.slice(2);
    for (let i = 0; i < argv.length; i++) {
        if (argv[i] === "--op") op = argv[++i];
        else if (argv[i] === "--cases-dir") casesDir = argv[++i];
        else if (argv[i] === "--logs-dir") logsDir = argv[++i];
    }

    const tomlPath = join(casesDir, `${op}.toml`);
    const doc = TOML.parse(readFileSync(tomlPath, "utf8"));
    const cfg = doc.config || {};
    const iterHint = cfg.iterations || 1000;
    const precision = cfg.precision || 28;
    const cases = doc.cases || [];

    if (!existsSync(logsDir)) mkdirSync(logsDir, { recursive: true });
    const stamp = ts();
    const logPath = join(logsDir, `js_${op}_${stamp}.csv`);

    console.log(`# js decimal.js ${op} (prec=${precision}, hint=${iterHint})`);
    console.log("case".padEnd(44) + "result".padEnd(36) + "ns/iter");

    const lines = ["timestamp,language,op,case_name,result,ns_per_iter"];
    for (const c of cases) {
        const name = c.name;
        const a = expand(c.a);
        const b = c.b ? expand(c.b) : "";
        let result, perNs;
        try {
            const [resStr, kernel] = makeKernel(op, a, b, precision);
            result = resStr;
            perNs = bench(kernel, iterHint);
        } catch (e) {
            result = `ERR: ${e.message}`;
            perNs = 0;
        }
        const short = result.length <= 34 ? result : result.slice(0, 34);
        console.log(name.padEnd(44) + short.padEnd(36) + perNs.toFixed(2));
        lines.push(
            [stamp, "js", op, csvQuote(name), csvQuote(result), perNs.toFixed(4)].join(",")
        );
    }
    writeFileSync(logPath, lines.join("\n") + "\n");
    console.error(`wrote ${logPath}`);
}

main();
